### PR TITLE
added the policy to check reference for secret in svcaccount

### LINKED ISSUE #819: [Sample] policies to audit and restrict long-lived tokens for service accounts
### Problem Statement

Prior to v1.24, Kubernetes generated long lived service accounts by default. These pose a security risk.

### Solution Description

Sample policies to:
* audit for secrets with `type: kubernetes.io/service-account-token`
* audit for service accounts with any `secrets.name` configured
* enforce pod `automountServiceAccountToken: false`

### Example "Good" Resource

_No response_

### Example "Bad" Resource

_No response_

### Other Comments

See:
- https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
- https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/
- https://eng.d2iq.com/blog/service-account-tokens-in-kubernetes-v1.24/

### Slack discussion

https://kubernetes.slack.com/archives/CLGR9BJU9/p1701451535229319

### Troubleshooting

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have verified the policy does not exist in the [samples library](https://kyverno.io/policies/).

### PR DESCRIPTION
**Pull request by swastik959**
## Related Issue(s)
partially addresses #819 

## Description
added the policy to check reference for secret in svcaccount 

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [x] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

### CODE REVIEW

**Comment by swastik959** [comment]
@chipzoller I have made a pr . though the kyverno tests were passing I am not sure why the chainsaw tests is failing . first when I ran it the error was . ``=== ERROR
        actual resource not found
`` and when I ran it second time the error has become ``  === STDERR
        Error from server (AlreadyExists): error when creating "STDIN": clusterpolicies.kyverno.io "check-service-accounts" already exists
``

**Comment by swastik959** [comment]
@chipzoller The ci for  imagepull policy is failing somehow even though it is unchanged

**Comment by swastik959** [comment]
@chipzoller please review this pr

File: other/check-service-account/check-service-account.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  background: true
+  rules:
+    - name: deny-secrets
+      match:
+        any:
+        - resources:
+            kinds:
+              - ServiceAccount
+      validate:
+        message: "long lived API tokens are not allowed"
```
Comment (chipzoller) [inline other/check-service-account/check-service-account.yaml:29]:
```suggestion
        message: "Long-lived API tokens are not allowed."
```

File: other/check-service-account/check-service-account.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    policies.kyverno.io/description: >-
+      Before version 1.24, Kubernetes automatically generated Secret-based tokens 
+      for ServiceAccounts. To distinguish between automatically generated tokens 
+      and manually created ones, Kubernetes checks for a reference from the 
+      ServiceAccount's secrets field. If the Secret is referenced in the secrets 
+      field, it is considered an auto-generated legacy token. These legacy Tokens can
+      be of security concern and should be audited.
+
+spec:
+  validationFailureAction: Enforce
```
Comment (chipzoller) [inline other/check-service-account/check-service-account.yaml:19]:
```suggestion
  validationFailureAction: Audit
```

File: other/check-service-account/check-service-account.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    policies.kyverno.io/severity: medium
+    policies.kyverno.io/subject: Secret,ServiceAccount
+    policies.kyverno.io/description: >-
+      Before version 1.24, Kubernetes automatically generated Secret-based tokens 
+      for ServiceAccounts. To distinguish between automatically generated tokens 
+      and manually created ones, Kubernetes checks for a reference from the 
+      ServiceAccount's secrets field. If the Secret is referenced in the secrets 
+      field, it is considered an auto-generated legacy token. These legacy Tokens can
+      be of security concern and should be audited.
+
```
Comment (chipzoller) [inline other/check-service-account/check-service-account.yaml:17]: Eliminate excess new line.

File: other/check-service-account/check-service-account.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      match:
+        any:
+        - resources:
+            kinds:
+              - ServiceAccount
+      validate:
+        message: "long lived API tokens are not allowed"
+        pattern:
+          X(secrets):
+
```
Comment (chipzoller) [inline other/check-service-account/check-service-account.yaml:32]: Eliminate excess new line. Only one new line at the end of a file.

File: other/check-service-account/check-service-account.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-service-accounts
+  annotations:
+    policies.kyverno.io/title: Check Exisitence of Secrets Reference in ServiceAccount
```
Comment (chipzoller) [inline other/check-service-account/check-service-account.yaml:6]:
```suggestion
    policies.kyverno.io/title: Check Long-Lived Secrets in ServiceAccounts
```

File: other/check-service-account/check-service-account.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    policies.kyverno.io/category: Sample
```
Comment (chipzoller) [inline other/check-service-account/check-service-account.yaml:7]:
```suggestion
    policies.kyverno.io/category: Security
```

**Review by chipzoller — CHANGES_REQUESTED** [review]
Missing the following annotations:

`kyverno.io/kubernetes-version`
`kyverno.io/kyverno-version`

**Comment by swastik959** [comment]
@chipzoller done

File: other/check-service-account/artifacthub-pkg.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      for ServiceAccounts. To distinguish between automatically generated tokens 
+      and manually created ones, Kubernetes checks for a reference from the 
+      ServiceAccount's secrets field. If the Secret is referenced in the secrets 
+      field, it is considered an auto-generated legacy token. These legacy Tokens can
+      be of security concern and should be audited.
+  
+      Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+annotations:
+  kyverno/category: "Sample"
+  kyverno/kubernetesVersion: "1.21"
```
Comment (chipzoller) [inline other/check-service-account/artifacthub-pkg.yaml:30]:
```suggestion
  kyverno/kubernetesVersion: "1.27"
```

File: other/check-service-account/artifacthub-pkg.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      Before version 1.24, Kubernetes automatically generated Secret-based tokens 
+      for ServiceAccounts. To distinguish between automatically generated tokens 
+      and manually created ones, Kubernetes checks for a reference from the 
+      ServiceAccount's secrets field. If the Secret is referenced in the secrets 
+      field, it is considered an auto-generated legacy token. These legacy Tokens can
+      be of security concern and should be audited.
+  
+      Refer to the documentation for more details on Kyverno annotations: https://artifacthub.io/docs/topics/annotations/kyverno/
+annotations:
+  kyverno/category: "Sample"
```
Comment (chipzoller) [inline other/check-service-account/artifacthub-pkg.yaml:29]:
```suggestion
  kyverno/category: "Security"
```

**Review by chipzoller — CHANGES_REQUESTED** [review]
I also think the title of this policy "check-service-accounts" (note how the name is accounts plural but account singular in the file name) is too generic and could be renamed to something more descriptive. This renaming would entail changing the folder name and all test files to match.

**Comment by swastik959** [comment]
@chipzoller any update on the pr ?

**Comment by chipzoller** [comment]
Please look through the current state of this PR to ensure that everything is correct and consistent, especially with respect to policy annotations and the Artifact Hub metadata file.

File: other/check-serviceaccount-secrets/artifacthub-pkg.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
+name: check-serviceaccount-secrets
+version: 1.0.0
+displayName: Check Existence of Secrets in ServiceAccount
+createdAt: "2024-01-06T20:14:33.000Z"
```
Comment (chipzoller) [inline other/check-serviceaccount-secrets/artifacthub-pkg.yaml:4]:
```suggestion
createdAt: "2024-03-02T06:14:33.000Z"
```

File: other/check-serviceaccount-secrets/artifacthub-pkg.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+description: >-
+  Before version 1.24, Kubernetes automatically generated Secret-based tokens 
+  for ServiceAccounts. To distinguish between automatically generated tokens 
+  and manually created ones, Kubernetes checks for a reference from the 
+  ServiceAccount's secrets field. If the Secret is referenced in the secrets 
+  field, it is considered an auto-generated legacy token. These legacy Tokens can
+  be of security concern and should be audited.
+install: |-
+  ```shell
+  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/check-service-account/check-service-account.yaml
```
Comment (chipzoller) [inline other/check-serviceaccount-secrets/artifacthub-pkg.yaml:14]:
```suggestion
  kubectl apply -f https://raw.githubusercontent.com/kyverno/policies/main/other/check-serviceaccount-secrets/check-serviceaccount-secrets.yaml
```

File: other/check-serviceaccount-secrets/.chainsaw-test/chainsaw-test.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+kind: Test
+metadata:
+  name: check-service-accounts
+spec:
+  steps:
+  - name: step-01
+    try:
+    - script:
+        content: |
+          sed 's/validationFailureAction: audit/validationFailureAction: Enforce/' ../check-serviceaccount-secrets.yaml | kubectl create -f -
```
Comment (chipzoller) [inline other/check-serviceaccount-secrets/.chainsaw-test/chainsaw-test.yaml:11]:
```suggestion
          sed 's/validationFailureAction: Audit/validationFailureAction: Enforce/' ../check-serviceaccount-secrets.yaml | kubectl create -f -
```

File: other/check-serviceaccount-secrets/.chainsaw-test/bad-svc-account.yaml [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+kind: ServiceAccount
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{},"name":"example-automated-thing","namespace":"examplens"}}      
+  name: bad-svc-account-02
+  namespace: default
+  resourceVersion: "777"
+  selfLink: /api/v1/namespaces/examplens/serviceaccounts/example-automated-thing
+  uid: f23fd170-66f2-4697-b049-e1e266b7f835
```
Comment (chipzoller) [inline other/check-serviceaccount-secrets/.chainsaw-test/bad-svc-account.yaml:11]:
```suggestion
```

File: other/check-serviceaccount-secrets/.chainsaw-test/bad-svc-account.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{},"name":"example-automated-thing","namespace":"examplens"}}
```
Comment (chipzoller) [inline other/check-serviceaccount-secrets/.chainsaw-test/bad-svc-account.yaml:6]:
```suggestion
```

**Review by chipzoller — APPROVED** [review]

**Pull request merged**